### ISSUE TITLE
problems with compiler.find_library()

### ISSUE THREAD

**Issue by rhd** [mesonbuild/meson#492]
Hi Jussi, I'm attempting to update my build files to support the latest meson code and I have run into several issues - I'm stuck on the last one.

- ~~`meson.get_compiler('cpp')` works worse than `meson.get_compiler('c')`~~
- A few code issues I found that you should add - I would have done a PR, but I got stuck on the issue below and need help
```diff
diff --git a/mesonbuild/interpreter.py b/mesonbuild/interpreter.py
index ef6f3c1..fd6b66e 100644
--- a/mesonbuild/interpreter.py
+++ b/mesonbuild/interpreter.py
@@ -762,12 +762,15 @@ class CompilerHolder(InterpreterObject):
         if not isinstance(required, bool):
             raise InterpreterException('required must be boolean.')
         search_dirs = kwargs.get('dirs', [])
+        if not isinstance(search_dirs, list):
+            search_dirs = [search_dirs]
         for i in search_dirs:
             if not os.path.isabs(i):
                 raise InvalidCode('Search directory %s is not an absolute path.' % i)
         linkargs = self.compiler.find_library(libname, search_dirs)
         if required and linkargs is None:
-            raise InterpreterException('Library %s not found'.format(libname))
+            raise InterpreterException('Library %s not found' % libname)
         lib = dependencies.ExternalLibrary(libname, linkargs)
         return ExternalLibraryHolder(lib)

```
- But this one I'm stuck on.
```
cc.find_library('mc_enc_avc', dirs: mydirs)
```
This command is getting executed by meson:
```
gcc-4.8.1 /tmp/tmpqd8qst99.c  -L/usr/local/dir1/ -L/usr/local/dir2 -lmc_enc_avc -o /tmp/tmpomiw489d
```
which outputs:
```
/usr/local/dir1/libmc_enc_avc.so: undefined reference to `dlsym'
/usr/local/dir1/libmc_enc_avc.so: undefined reference to `dlerror'
/usr/local/dir1/libmc_enc_avc.so: undefined reference to `pthread_create'
/usr/local/dir1/libmc_enc_avc.so: undefined reference to `pthread_mutexattr_init'
/usr/local/dir1/libmc_enc_avc.so: undefined reference to `dlopen'
/usr/local/dir1/libmc_enc_avc.so: undefined reference to `dlclose'
/usr/local/dir1/libmc_enc_avc.so: undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
```

The link test never succeeds even though the compiler can find the library because that library needs `pthread`.

Thoughts on this?

edit: I can't reproduce the first issue
edit2: adding `-lpthread -ldl` allows it to work. 
edit3: passing this on the gcc line allows the test to pass: `-Wl,--unresolved-symbols=ignore-in-shared-libs`

**Comment by rhd** [comment]
Another issue I'm having is this:

```
$ find /usr/ -name libfftw3.*
find: `/usr/lib64/audit': Permission denied
/usr/lib64/libfftw3.so.3.2.3
/usr/lib64/libfftw3.so.3
/usr/lib64/libfftw3.so
/usr/local/fftw/32/lib/libfftw3.la
/usr/local/fftw/32/lib/libfftw3.a
/usr/local/fftw/64/lib/libfftw3.la
/usr/local/fftw/64/lib/libfftw3.a
```

When I call `cc.find_library` I give it the `/usr/local` path. But when the `gcc` command line gets built, it does this: `'-L/usr/local/fftw/64/lib' '-lfftw3f'` but it actually picks up the wrong library - the one ine `/usr/lib64`

The obvious solution is to fix the crappy dev box, but this worked great with the old standalone version of `find_library()`.

But what this means is that it's not possible to test newer versions of libraries without removing the old ones first. I prefer the old behavior.

**Comment by jpakkane** [comment]
The new one has several issues and we might need to change it. However in this case what you want to do is to use pkg-config. Fftw3 ships one so just use that. Never, ever, ever use find_library if you can avoid it, even if done perfectly it is still an unreliable horror show.

**Comment by rhd** [comment]
I can't avoid find_library on these machines. People just rsync /usr/local/ from box to box. It sucks. But it looks like I can't update to the new `find_library` yet. :/

**Comment by rhd** [comment]
Is there any way you can un-deprecate the standalone `find_library()` until everyone is happy with the new compiler specific version?

**Comment by rhd** [comment]
fixed in #501 - thanks a lot!
